chain/network/src/network_protocol/network.proto

```diff
@@ -5,6 +5,22 @@ package network;
 
 import "google/protobuf/timestamp.proto";
 
+// Proof that a given peer owns the account key.
+// Included in every outbound TIER1 handshake.
+// Note: sign AccountKeyPayload, rather than OwnedAccount directly.
+message OwnedAccount {
+  PublicKey account_key = 1; // required
+  // PeerId of the node owning the account_key.
+  PublicKey peer_id = 2; // required 
+  // Timestamp indicates the date of signing - we do not assume the
```

Comment:
Is this timestamp used to prevent replay attacks? Because if it is, I wouldn't feel it's very effective. Especially if we're sending the signed OwnedAccount through the tier2 handshake too. It would be easy for someone to get a hold of a valid signature and then replay it to another validator, to join tier 1.

Standard asymmetric encryption would not help unless the encryption uses the same key as the validator's signing key - which may or may not be easy to do, I'm not familiar with the code.

A three-way handshake with a random challenge to sign should be done with a well-established handshake algorithm, to avoid being tricked to sign something that is not supposed to be signed, like a transaction :)

Or, could just use standard asymmetric encryption handshake, and then send a message signing the public key used to make the handshake. That will make it not replayable since nobody else can prove ownership of that public key through the handshake, and at the same time we can be sure that the validator key is only ever used to sign another public key (that we generate) and nothing else.

Reply:
> I wouldn't feel it's very effective.

True. It is not very effective, I'm aware of that. This was supposed to be a short term poor man's solution until I implement the 3-way handshake which will require another refactor of PeerActor (3-way handshake will increase the number of states of PeerActor state machine). If you think it would be beneficial to do 3-way handshake from day 1, I can invest time into that.

> A three-way handshake with a random challenge to sign should be done with a well-established handshake algorithm

Yes, I'd prefer to not implement the 3-way handshake by hand, but rather use some existing solution (just so that I don't have to maintain it, if not for anything else). Moreover I'd love to have all our traffic encrypted e2e just for defense in depth. Do you have any suggestions which library could we utilize? I guess some rust-native library would be preferred.

> to avoid being tricked to sign something that is not supposed to be signed, like a transaction

This is the analysis I did of payloads that are signed with the validator key:
https://docs.google.com/document/d/13xxzpEl7x0KwEKiu5bt2y1Rx0GjEY2A8a72uXeUPWmY/edit#heading=h.3t0iopefbw2u

As you can see, the stuff that is being signed is already poorly structured.
The way going forward, as I see it, would be to have an enum of all the signable things to prevent collisions.
For now I've introduced such an enum here: https://cs.github.com/near/nearcore/blob/37232dbc9af61cf60d1ed0b7248ddd8f683d5110/chain/network/src/network_protocol/network.proto#L18
and I hope that eventually we can put everything into that enum, but right now all guarantee we have is that doc of mine I guess.

> Or, could just use standard asymmetric encryption handshake, and then send a message signing the public key used to make the handshake.

In a sense we already do sth like that: we sign PeerId which is also a public key, and we additionally sign an edge which contains public keys of both ends. I don't really want to rely on that, but it kind of prevents the replay attacks in my poor man's solution, right?